src/scf/newton.jl

```diff
@@ -27,100 +35,76 @@
-#   - computing Kδψ can be done by constructing the dρ associated with dψ, then
-#     using the exchange-correlation kernel to compute the associated dV, then
-#     acting with dV on the ψ and projecting on the tangent space;
-#   - to perform Newton iterations, we solve (Ω+K)δψ = Hψ-λψ and then perform
-#     ψ = ψ - δψ, with proper orthonormalization.
-#
-# For further details :
-# Eric Cancès, Gaspard Kemlin, Antoine Levitt. Convergence analysis of
-# direct minimization and self-consistent iterations. SIAM Journal of Matrix
-# Analysis and Applications, 42(1):243–274 (2021)
+#     we chose this one because it makes Ω self-adjoint for the Frobenius scalar
+#     product;
+#   - computing Kδψ can be done by constructing the δρ associated with δψ, then
+#     using the exchange-correlation kernel to compute the associated δV, then
+#     acting with δV on the ψ and projecting on the tangent space.
 
 using LinearMaps
 using IterativeSolvers
 
-# Tools
-
-"""
-    compute_projected_gradient(basis::PlaneWaveBasis, ψ, occupation)
-
-Compute the residual associated to a set of planewave ψ, that is to say
-H(ψ)*ψ - ψ*λ where λ is the set of Rayleigh coefficients associated to the ψ.
-"""
+#  Compute the gradient of the energy, projected on the space tangent to ψ, that
+#  is to say H(ψ)*ψ - ψ*λ where λ is the set of Rayleigh coefficients associated
+#  to the ψ.
 function compute_projected_gradient(basis::PlaneWaveBasis, ψ, occupation)
-    Nk = length(basis.kpoints)
     ρ = compute_density(basis, ψ, occupation)
     _, H = energy_hamiltonian(basis, ψ, occupation; ρ=ρ)
 
-    [proj_tangent_kpt(H.blocks[ik]*ψ[ik], ψ[ik]) for ik = 1:Nk]
+    [proj_tangent_kpt(H.blocks[ik] * ψ[ik], ψ[ik])
+     for ik = 1:length(basis.kpoints)]
 end
 
-"""
-    proj_tangent_kpt(δψk, ψk)
-
-Computation the projection of δψk onto the tangent space defined at ψk, where
-δψk and ψk are defined at the same kpoint.
-"""
-proj_tangent_kpt(δψk, ψk) = δψk - ψk * (ψk'δψk)
+# Projections on the space tangent to ψ
 function proj_tangent_kpt!(δψk, ψk)
+    # δψk <- δψk - ψk * (ψk'δψk)
     mul!(δψk, ψk, ψk'δψk, -1, 1)
 end
-"""
-    proj_tangent(δψ, ψ)
-
-Computation the projection of δψ onto the tangent space defined at ψ.
-"""
+proj_tangent_kpt(δψk, ψk) = proj_tangent_kpt!(copy(δψk), ψk)
 function proj_tangent(δψ, ψ)
     [proj_tangent_kpt(δψ[ik], ψ[ik]) for ik = 1:size(ψ,1)]
 end
 function proj_tangent!(δψ, ψ)
     [proj_tangent_kpt!(δψ[ik], ψ[ik]) for ik = 1:size(ψ,1)]
 end
 
-# Operators
+function apply_Ω(basis::PlaneWaveBasis, δψ, ψ, H::Hamiltonian)
 
-"""
-    apply_Ω(basis::PlaneWaveBasis, δψ, ψ, H)
+    δψ = proj_tangent(δψ, ψ)
+    Ωδψ = map(1:length(basis.kpoints)) do ik
+        ψk = ψ[ik]
+        Hk = H.blocks[ik]
+        δψk = δψ[ik]
 
-Apply Ω to an element in the tangent space to ψ.
-"""
-function apply_Ω(basis::PlaneWaveBasis, δψ, ψ, H)
-    Nk = length(basis.kpoints)
+        Hψk = Hk * ψk
+        ψHψk = Hermitian(ψk * Hψk')
```

Comment:
That object is just weird, so don't give it a name and just write the thing inline

Reply:
... Actually, don't. This is a huge object, so never compute it explicitly.

Reply:
I'm not sure what to do then ?

Reply:
(XY')Z is bad, X(Y'Z) is good